[None][feat] Add Qwen3 MoE support to TensorRT backend

gkswns0531 opened the pull request:
Implements TensorRT-LLM support for Qwen3 MoE models with architectural differences from Qwen2 MoE:

## Implement
- Add 'Qwen3MoeForCausalLM' to MODEL_MAP for automatic detection
- Add 'qwen3_moe' to valid model types in QWenConfig
- Use standard MOE implementation for Qwen3 MoE **without shared experts**
- Update weight conversion logic to handle Qwen3 MoE's expert structure
- Refactor trtllm_modules_to_hf_modules mapping to distinguish between Qwen2 MoE (with shared experts) and Qwen3 MoE (without shared experts)


## Testing
- Successfully creates Qwen3-30B-A3B MoE Engine (FP16, FP8)

ref)
vllm vs tensorrt-llm latency (Qwen3-30B-A3B MoE, H100 PCIe GPU, 2048 input_length, 3072 max_length, batch 16)

<html>
<head>

</head>
<body>

Precision | Framework | TTFT (ms) | TPS (tokens/s)
-- | -- | -- | --
FP16 | vLLM | 152.750 | 727.76
  | TRT (torch backend) | 127.477 | 785.82
  | TRT (trt backend) | 73.729 | 764.59
FP8 | vLLM | 175.951 | 993.81
  | TRT (torch backend) | 128.124 | 1164.08
  | TRT (trt backend) | 84.660 | 1206.60


</body>
</html>

Dear Maintainers,

I would like to kindly submit a pull request that adds support for building the Qwen3-MoE model as a TensorRT engine.
I would be truly grateful if you could take the time to review it at your convenience.

Thank you very much for your consideration.



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

* **New Features**
  * Added support for the "qwen3_moe" model variant, including activation functions, model mappings, and weight loading.
  * Introduced new configuration options for layer-specific MLP control and decoder sparsity steps.
* **Refactor**
  * Enhanced conditional logic for handling Mixture of Experts (MoE) models, improving clarity and extensibility.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

gkswns0531 commented:
@Dramilio 
I have successfully built the Qwen3 MoE engine (FP16, FP8). Please pull the latest PR and use it.
I plan to request reviews from the reviewers during the weekdays.
Thanks!

gkswns0531 commented:
@byshiue @nv-yilinf @shaharmor98 
This is the PR for building the Qwen3 MoE engine. I’ve completed the testing, so when you have some time, I would greatly appreciate your review.

Thank you!

byshiue commented:
/bot run

tensorrt-cicd commented:
[PR_Github #13895](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13895/) [ run ] triggered by Bot

gkswns0531 commented:
/bot run

byshiue commented:
/bot run

tensorrt-cicd commented:
[PR_Github #13969](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13969/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #13969](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13969/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #10522](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10522/) completed with status: 'FAILURE'

gkswns0531 committed:
Add Qwen3 MoE support to TensorRT-LLM backend

Signed-off-by: gkswns0531 <gkswns0531@gmail.com>

gkswns0531 committed:
Fix code formatting issues

Signed-off-by: gkswns0531 <gkswns0531@gmail.com>

gkswns0531 committed:
feat: Add Qwen3 MoE support

Signed-off-by: hanjuncho <gkswns0531@gmail.com>
Signed-off-by: gkswns0531 <gkswns0531@gmail.com>

gkswns0531 committed:
Improve Qwen3 MoE implementation following industry standard

- Follow the same pattern as DeepSeek V2 and LLaMA/Mixtral MoE models
- Remove unnecessary mlp.gate and mlp.router mappings
- Use existing processed MoE weights instead of arbitrary first expert selection
- Only create mlp.fc and mlp.proj for quantization compatibility

Signed-off-by: gkswns0531 <gkswns0531@gmail.com>

gkswns0531 committed:
Fix Qwen3 MoE quantization error and add missing config support

- Add Qwen3MoeForCausalLM to model registry
- Fix MLP block mapping for quantization compatibility
- Add mlp_only_layers and decoder_sparse_step config support
- Preserve existing Qwen2 MoE functionality

Signed-off-by: gkswns0531 <gkswns0531@gmail.com>

gkswns0531 committed:
Fix Qwen3 MoE weight shape mismatch in convert.py

Remove incorrect quantization compatibility code that was creating
duplicate mlp.fc weights with wrong shapes for Qwen3 MoE models.
Qwen3 MoE only needs individual expert weights, not aggregated ones.

Signed-off-by: gkswns0531 <gkswns0531@gmail.com>

gkswns0531 committed:
Fix code formatting issues for Qwen3 MoE implementation

Signed-off-by: gkswns0531 <gkswns0531@gmail.com>

gkswns0531 commented:
Oops, sorry about that. I changed my environment recently, and that caused a DCO sign-off mismatch. I’ve fixed it and pushed again with a consistent sign-off.

byshiue commented:
/bot run

tensorrt-cicd commented:
[PR_Github #14053](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14053/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #14053](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14053/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #10601](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10601/) completed with status: 'FAILURE'

byshiue committed:
Merge branch 'main' into moe-qwen3-updated

tensorrt-cicd commented:
[PR_Github #14063](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14063/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10609](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10609/) completed with status: 'FAILURE'

byshiue commented:
/bot run

tensorrt-cicd commented:
[PR_Github #14095](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14095/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #14095](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14095/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10635](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10635/) completed with status: 'FAILURE'

byshiue reviewed (approved):
LGTM

vip-china commented:
I tried the quantized versions of fp8 and int4 here, and there were instances of duplicate answers. I used convert_checkpoint.by and quantize.by respectively @gkswns0531 @Dramilio 
<img width="908" height="893" alt="d0da96e35b8136038c61bb776b64f21" src="https://github.com/user-attachments/assets/c653fe99-7a98-4462-9eb6-d3e6c21c0539" />

gkswns0531 commented:
@vip-china 

Hello,
Could you please share the exact command you used, the TensorRT-LLM version, and—if you are using Triton—the Triton version as well?
It would also be very helpful if you could provide more detailed error logs so that we can take a closer look.
Thank you.

Smallcuteyu commented:
Hello? any update @gkswns0531  @byshiue

gkswns0531 commented:
@Smallcuteyu @vip-china 

Thank you for reporting the issue. However, my project is in a very busy season until early October, so I will check it after that.

Thanks
